### PR TITLE
[Feature] Make project temporal

### PR DESCRIPTION
**Pull request by YasheshSavani**
Description:

This PR will add a toggle button on project details page which will be used to make the project temporal or not. And also, the user choice will be saved in the project config for future use.

Update:
- This PR will replace methods created for `use_gpu` and with more generic implementation which can be used by both `use_gpu` and `project_temporal` property as well as in future implementation of toggle buttons. 
- By default, it will also add two properties `use_gpu` and `project_temporal` with value `True` when a new project is being created.

### CODE REVIEW

File: tools/sense_studio/sense_studio.py [hunk 1]
```diff
@@ -206,12 +208,25 @@ def toggle_gpu():
     """
     data = request.json
     path = data['path']
-
-    use_gpu = utils.toggle_gpu_status(path)
+    button_name = data['button_name']
```
Comment (YasheshSavani) [inline tools/sense_studio/sense_studio.py:211]: Thinking on a better name something like `button_type` or `property_name`. Do suggest if anything looks good @corneliusboehm @manikd31

File: tools/sense_studio/templates/project_details.html [hunk 2]
```diff
@@ -12,10 +12,18 @@
         <h3 class="ui header center aligned">Settings</h3>
         <div class="ui toggle checkbox hashoverpopup"
            data-content="Should GPU be used for computing features?"
-           onclick="toggleGPU('{{ path }}');">
+           onclick="toggleGPU('{{ path }}', 'use_gpu');">
```
Comment (YasheshSavani) [inline tools/sense_studio/templates/project_details.html:15]: Her in my view, we can make a constant variable for `'use_gpu'` and `'project_temporal'` strings.

File: tools/sense_studio/utils.py [hunk 3]
```diff
@@ -22,7 +22,7 @@ def load_feature_extractor(project_path):
     feature_extractor.eval()
 
     # Create Inference Engine
-    use_gpu = get_gpu_status(project_path)
+    use_gpu = get_toggle_button_status(project_path, 'use_gpu')
```
Comment (YasheshSavani) [inline tools/sense_studio/utils.py:25]: That constant can be used here for `'use-gpu'`.

File: tools/sense_studio/static/main.js [hunk 4]
```diff
@@ -226,13 +226,48 @@ function editClass(index, shouldEdit) {
...
+    } else {
+        makeProjectTemporal.removeAttribute('checked');
+
+        // Hide tags on project details page
+        for (let i=0; i < classTagsDiv.length; i++){
+            classTagsDiv[i].style.display = 'none';
+            editClassTagsDiv[i].style.display = 'none';
+        }
+        addClassTagsDiv.style.display = 'none';
+    }
```
Comment (manikd31) [inline tools/sense_studio/static/main.js:268]: We can shorten the lines of code by doing something like this:
```
let displayStyle = 'none';

if (response.project_temporal) {
    makeProjectTemporal.setAttribute('checked', 'checked');
    displayStyle = 'block';
} else {
    makeProjectTemporal.removeAttribute('checked');
}

for (let i = 0; i < classTagsDiv.length; i++) {
    classTagsDiv[i].style.display = displayStyle;
    editClassTagsDiv[i].style.display = displayStyle;
}
addClassTagsDiv.style.display = displayStyle;
```

File: tools/sense_studio/templates/project_details.html [hunk 5]
```diff
@@ -12,10 +12,18 @@
...
             <input type="checkbox" id="gpuInput" {% if config.use_gpu %} checked="checked" {% endif %}>
             <label>Use GPU</label>
         </div>
+        <br/>
+        <br/>
+        <div class="ui toggle checkbox hashoverpopup"
+           data-content="Do you want to make project temporal?"
+           onclick="toggleMakeProjectTemporal('{{ path }}', '{{ project }}', 'project_temporal');">
+            <input type="checkbox" id="makeProjectTemporal" {% if config.project_temporal %} checked="checked" {% endif %}>
+            <label>Make Project Temporal</label>
```
Comment (manikd31) [inline tools/sense_studio/templates/project_details.html:25]: Instead of a long string "Make Project Temporal", can we rename it to "Temporal Training"?

**Comment by manikd31** [comment]
Disabling the `temporal training` toggle button hides the respective divs, but the button for `Annotate` is still visible. Also, we might need to change the text in each class which shows: `Train: 5 (0 annotated)`.

File: tools/sense_studio/static/main.js (hunk 4 above)
Comment (YasheshSavani) [reply tools/sense_studio/static/main.js:268]: Thanks

File: tools/sense_studio/templates/project_details.html (hunk 5 above)
Comment (YasheshSavani) [reply tools/sense_studio/templates/project_details.html:25]: we can let's think of a name which is better understood I think.

File: tools/sense_studio/templates/project_details.html [hunk 6]
```diff
@@ -171,7 +183,7 @@ <h4>Tags</h4>
                     </div>
                 </div>
             </div>
-            <div class="content">
+            <div class="content addClassTags" {% if not config.project_temporal %} style="display: none;" {% endif %} >
```
Comment (corneliusboehm) [inline tools/sense_studio/templates/project_details.html:186]: Please make this a CSS class

File: tools/sense_studio/utils.py [hunk 7]
```diff
@@ -90,17 +90,19 @@ def get_class_labels(path):
     return config['classes'].keys()
 
 
-def get_gpu_status(path):
+def get_toggle_button_status(path, button_name):
```
Comment (corneliusboehm) [inline tools/sense_studio/utils.py:93]: I would propose calling this `get_project_setting` instead, because toggle button is very UI-related.

**Review by corneliusboehm — COMMENTED** [review]
I like your idea of making the functions for getting and setting project settings generic :+1: 
Good feedback from @manikd31 already and I've added a few more comments.

**Comment by YasheshSavani** [comment]
Still working on annotation button hiding on navbar as it's a bit tricky.

File: tools/sense_studio/static/main.css [hunk 8]
```diff
@@ -70,3 +70,12 @@ a {
...
 }
+
+.displayTags {
+    display: none;
+}
+
+.hiddenButton {
+    visibility:hidden;
+    display:none;
+}
```
Comment (manikd31) [inline tools/sense_studio/static/main.css:81]:
```suggestion
    visibility: hidden;
    display: none;
}

```

File: tools/sense_studio/static/main.js [hunk 9]
```diff
@@ -226,13 +226,51 @@ function editClass(index, shouldEdit) {
...
+    for (let i=0; i < classTagsDiv.length; i++){
+        classTagsDiv[i].style.display = displayStyle;
+        editClassTagsDiv[i].style.display = displayStyle;
+    }
+    addClassTagsDiv[0].style.display = displayStyle;
+
+    // Redirect back to project details page (It will refresh the page).
+//    location.href = '/project/' + project;
+
+}
```
Comment (manikd31) [inline tools/sense_studio/static/main.js:276]:
```suggestion
}

```

File: tools/sense_studio/templates/navigation.html [hunk 10]
```diff
@@ -26,7 +26,7 @@
     </div>
 </div>
 <div class="ui simple dropdown item">
-    Annotation
+        Annotation
```
Comment (manikd31) [inline tools/sense_studio/templates/navigation.html:29]:
```suggestion
    Annotation
```

File: tools/sense_studio/templates/project_details.html [hunk 11]
```diff
@@ -12,10 +12,18 @@
...
            data-content="Should GPU be used for computing features?"
-           onclick="toggleGPU('{{ path }}');">
+           onclick="toggleGPU('{{ path }}', 'use_gpu');">
             <input type="checkbox" id="gpuInput" {% if config.use_gpu %} checked="checked" {% endif %}>
             <label>Use GPU</label>
         </div>
+        <br/>
+        <br/>
+        <div class="ui toggle checkbox hashoverpopup"
+           data-content="Do you want to make project temporal?"
```
Comment (manikd31) [inline tools/sense_studio/templates/project_details.html:22]:
```suggestion
           data-content="Do you want to use temporal annotations?"
```

**Review by corneliusboehm — APPROVED** [review]
Same as last time: I made some updates and simplifications, so let me know if you want to talk about those.
Everything works fine now and we can merge in my opinion :+1:

**Comment by YasheshSavani** [comment]
> Same as last time: I made some updates and simplifications, so let me know if you want to talk about those.
> Everything works fine now and we can merge in my opinion +1

Changes look good to me too. Thanks for the updates :+1: .

**Pull request merged**